cloud/terraform/machineactuator.go

```diff
@@ -351,10 +353,119 @@ func (tf *TerraformClient) PostDelete(cluster *clusterv1.Cluster, machines []*cl
+	// Try to get the annotations for the versions. If they don't exist, update the annotation and return.
+	// This can only happen right after bootstrapping.
+	if goalMachine.ObjectMeta.Annotations == nil {
+		ip, _ := tf.GetIP(goalMachine)
+		glog.Info("Populating existing state for bootstrapped machine.")
+		return tf.updateAnnotations(goalMachine, ip)
+	}
+
+	if util.IsMaster(goalMachine) {
+		glog.Info("Upgrade for master machine.. Check if upgrade needed.")
+		// Get the versions for the new object.
+		goalMachineControlPlaneVersion := goalMachine.Spec.Versions.ControlPlane
+		goalMachineKubeletVersion := goalMachine.Spec.Versions.Kubelet
+
+		// If the saved versions and new versions differ, do in-place upgrade.
+		if goalMachineControlPlaneVersion != goalMachine.Annotations[ControlPlaneVersionAnnotationKey] ||
+			goalMachineKubeletVersion != goalMachine.Annotations[KubeletVersionAnnotationKey] {
+			glog.Infof("Doing in-place upgrade for master from %s to %s", goalMachine.Annotations[ControlPlaneVersionAnnotationKey], goalMachineControlPlaneVersion)
+			err := tf.updateMasterInPlace(goalMachine)
+			if err != nil {
+				glog.Errorf("Master inplace update failed: %+v", err)
+			}
+		} else {
+			glog.Info("UPDATE TYPE NOT SUPPORTED")
+			return nil
+		}
+	} else {
+		glog.Info("NODE UPDATES NOT SUPPORTED")
+		return nil
+	}
+
 	return nil
 }
 
+// Assumes that update is needed.
+// For now support only K8s control plane upgrades.
+func (tf *TerraformClient) updateMasterInPlace(goalMachine *clusterv1.Machine) error {
+	goalMachineControlPlaneVersion := goalMachine.Spec.Versions.ControlPlane
+	goalMachineKubeletVersion := goalMachine.Spec.Versions.Kubelet
+
+	currentControlPlaneVersion := goalMachine.Annotations[ControlPlaneVersionAnnotationKey]
+	currentKubeletVersion := goalMachine.Annotations[KubeletVersionAnnotationKey]
+
+	// Control plane upgrade
+	if goalMachineControlPlaneVersion != currentControlPlaneVersion {
+		// Pull the latest kudeadm
+		cmd := "export KUBEADM_VERSION=$(curl -sSL https://dl.k8s.io/release/stable.txt); " +
+			"curl -sSL https://dl.k8s.io/release/${KUBEADM_VERSION}/bin/linux/amd64/kubeadm | sudo tee /usr/bin/kubeadm > /dev/null; " +
+			"sudo chmod a+rx /usr/bin/kubeadm"
+		_, err := tf.remoteSshCommand(goalMachine, cmd, "~/.ssh/id_rsa", "ubuntu")
+		if err != nil {
+			glog.Infof("remoteSshCommand failed while downloading new kubeadm: %+v", err)
+			return err
+		}
+
+		// Next upgrade control plane
+		cmd = fmt.Sprintf("sudo kubeadm upgrade apply %s -y", "v"+goalMachine.Spec.Versions.ControlPlane)
+		_, err = tf.remoteSshCommand(goalMachine, cmd, "~/.ssh/id_rsa", "ubuntu")
+		if err != nil {
+			glog.Infof("remoteSshCommand failed while upgrading control plane: %+v", err)
+			return err
+		}
+	}
+
+	// Upgrade kubelet
+	if goalMachineKubeletVersion != currentKubeletVersion {
+		cmd := fmt.Sprintf("sudo kubectl drain %s --kubeconfig /etc/kubernetes/admin.conf --ignore-daemonsets", goalMachine.Name)
```

Comment:
Wouldn't this evict the machine controller?

Reply:
Yes I tested this and turns out the controller should not be evicted.